dsf/org.eclipse.cdt.dsf.ui/src/org/eclipse/cdt/dsf/debug/ui/actions/DsfSteppingModeTarget.java

```diff
@@ -76,15 +76,15 @@ public boolean supportsInstructionStepping() {
 	 */
 	@Override
 	public void addPropertyChangeListener(IPropertyChangeListener listener) {
-		fPreferences.addPropertyChangeListener(listener);
+		// Left unimplemented because of backward compatibility.
```

Comment:
If you want to rework access to preferences in this PR please provide a redirection for adding/removing property change listeners, otherwise API clients who need to listen preferences will be broken.

Alternative you can postpone preference rework for another PR @raghucssit

Reply:
This can be deferred. I will fix as a separate PR if there is any client affected.
The reason is: New implementation uses `org.eclipse.core.runtime.preferences.IEclipsePreferences` which can only accept listeners of type `org.eclipse.core.runtime.preferences.IEclipsePreferences.IPreferenceChangeListener`.
**Option 1**:  we need to maintain a map of all the listeners (`IPropertyChangeListener` --> `IPreferenceChangeListener`) in `DsfSteppingModeTarget` and pass the preference change event to `IPropertyChangeListener`.
**Option 2**: We can introduce new API at `org.eclipse.cdt.debug.core.model.ITargetProperties` which accepts `IPreferenceChangeListener`.
So I am planning to fix this only if needed in future.

Reply:
How will you be able to identify if this is needed or not?
The problem here (and the concern) is that the ITargetProperties is a interface that is part of the API and has certain expectations from the consumer. Today, the contract for the interface is that it should be possible to register a listener to get notifications on changes. If someone uses this or not is not really important.

By implementing the add/remove listener methods with an empty body breaks these assumptions from anyone trying to use the ITargetProperties interface by trying to register a listener.

With all this said, I think the only viable option is to proxy/transform the events between the different listener interfaces to keep the API intact.
Note: Option 2 above would not solve anything as it would actually cause another breakage as an already existing API interface cannot simply be extended without causing issues for anyone implementing the interface.

Reply:
> **Option 1**: we need to maintain a map of all the listeners (`IPropertyChangeListener` --> `IPreferenceChangeListener`) in `DsfSteppingModeTarget` and pass the preference change event to `IPropertyChangeListener`.

This approach is also not possible because `PropertyChangeEvent(Object source, String property, Object oldValue, Object newValue)` has to carry back the preference object itself as `source`. With new preference used we don't have old preference object. And constructing an old preference object from new preference object is difficult both have significantly different structure. 
So I have changed the code to use old preference approach.
i.e. Now there is no change needed for `DsfSteppingModeTarget`.

Reply:
@Torbjorn-Svensson I have made the above fix. Please run the build on the latest commit.